src/controllers/userController.js

```diff
@@ -0,0 +1,163 @@
+const admin = require('../config/firebase');
+const db = admin.firestore();
+const axios = require('axios');
+
+exports.signupUser = async (req, res) => {
+    // console.log(req.body)
+    const { name, email, password, IsAdmin } = req.body;
+
+    if (!name || !email || !password) {
+        return res.status(400).json({ error: 'All fields are required' });
+    }
+
+    try {
+        const userRecord = await admin.auth().createUser({
+            displayName: name,
+            email,
+            password
+        });
+        // console.log('User created:', userRecord);
+
+        if (IsAdmin) {
+            await admin.auth().setCustomUserClaims(userRecord.uid, { admin: true });
+            // console.log('Admin claim set for user:', userRecord.uid);
+        }
+
+        await db.collection('users').doc(userRecord.uid).set({
+            name,
+            email,
+            IsAdmin,
+            createdAt: admin.firestore.FieldValue.serverTimestamp(),
+        });
+        // console.log('User details saved in Firestore:', userRecord.uid);
+
+        res.status(201).send({ message: 'User signed up successfully', user: userRecord });
+    } catch (error) {
+        console.error('Error signing up user:', error);
+        res.status(400).send({ error: error.message });
+    }
+};
+
+exports.loginUser = async (req, res) => {
+    const { email, password } = req.body;
+
+    try {
+        // console.log('Attempting to log in with email:', email);
+        const apiKey = process.env.FIREBASE_API_KEY;
+        // console.log('Firebase API Key:', apiKey);
+
+        const response = await axios.post(`https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=${apiKey}`, {
+            email,
+            password,
+            returnSecureToken: true
+        });
+
+        const idToken = response.data.idToken;
+        const userRecord = await admin.auth().getUserByEmail(email);
+        // console.log('ID token generated for user:', userRecord.uid);
+
+        res.status(200).send({ message: 'User logged in successfully', token: idToken, user: userRecord });
+    } catch (error) {
+        console.error('Error logging in user:', error);
+        if (error.response) {
+            console.error('Error response data:', error.response.data); // Log the error response data
+        }
+        res.status(400).send({ error: error.message });
+    }
+};
+
+exports.getUser = async (req, res) => {
+    const { uid } = req.params;
+
+    try {
+        const userRecord = await admin.auth().getUser(uid);
+        res.status(200).send(userRecord);
+    } catch (error) {
+        console.error('Error fetching user:', error);
+        res.status(400).send({ error: error.message });
+    }
+};
+
+exports.getAllUsers = async (req, res) => {
+    try {
+        const usersSnapshot = await db.collection('users').get();
+        if (usersSnapshot.empty) {
+            return res.status(404).send({ message: 'No users found' });
```

Comment:
I think this should just return an empty list and a `200` status. We should only return a 404 if we request a specific user that doesn't exist (which we don't do currently and should fix).